zkevm-circuits/src/evm_circuit/execution/error_oog_codestore.rs

```diff
@@ -0,0 +1,221 @@
+        cb.require_zero("in internal call", cb.curr.state.is_root.expr());
+
+        // Case C in the return specs.
+        let restore_context = RestoreContextGadget::construct(
+            cb,
+            0.expr(),
+            0.expr(),
+            memory_address.offset(),
+            memory_address.length(),
+            0.expr(),
+            0.expr(),
+        );
+
+        Self {
+            opcode,
+            is_create,
+            memory_address,
+            code_store_gas_insufficient,
+            restore_context,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = step.opcode.unwrap();
+        self.opcode
+            .assign(region, offset, Value::known(F::from(opcode.as_u64())))?;
+
+        let [memory_offset, length] = [0, 1].map(|i| block.rws[step.rw_indices[i]].stack_value());
+        self.memory_address
+            .assign(region, offset, memory_offset, length)?;
+
+        self.is_create
+            .assign(region, offset, Value::known(F::from(call.is_create as u64)))?;
+        self.code_store_gas_insufficient.assign(
+            region,
+            offset,
+            F::from(step.gas_left),
+            F::from(200 * length.as_u64()),
+        )?;
+        self.restore_context
+            .assign(region, offset, block, call, step, 3)?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use bus_mapping::circuit_input_builder::CircuitsParams;
+    use eth_types::{
+        address, bytecode, evm_types::OpcodeId, geth_types::Account, Address, Bytecode, Word,
+    };
+
+    use lazy_static::lazy_static;
+    use mock::{eth, TestContext};
+
+    use crate::test_util::CircuitTestBuilder;
+
+    const CALLEE_ADDRESS: Address = Address::repeat_byte(0xff);
+    lazy_static! {
+        static ref CALLER_ADDRESS: Address = address!("0x00bbccddee000000000000000000000000002400");
+    }
+
+    fn run_test_circuits(ctx: TestContext<2, 1>) {
+        CircuitTestBuilder::new_from_test_ctx(ctx)
+            .params(CircuitsParams {
+                max_rws: 4500,
+                ..Default::default()
+            })
+            .run();
+    }
+
+    // RETURN or REVERT with data of [0x60; 5]
+    fn initialization_bytecode() -> Bytecode {
+        let memory_bytes = [0x60; 10];
+        let memory_address = 0;
+        let memory_value = Word::from_big_endian(&memory_bytes);
+        let mut code = bytecode! {
+            PUSH10(memory_value)
+            PUSH1(memory_address)
+            MSTORE
+            PUSH2(5) // length to copy
+            PUSH2(32u64 - u64::try_from(memory_bytes.len()).unwrap())
+        };
+        code.write_op(OpcodeId::RETURN);
+
+        code
+    }
+
+    fn creater_bytecode(initialization_bytecode: Bytecode, is_create2: bool) -> Bytecode {
```

Comment:
creator

Reply:
updated in [fbea334](https://github.com/scroll-tech/zkevm-circuits/pull/304/commits/fbea3347b5e799723a89c9a3c8e2b5eb05579c22) thanks